module/VuFind/src/VuFind/Db/Service/TagService.php

```diff
@@ -778,4 +780,399 @@ public function getStatistics(bool $extended = false): array
         }
         return $stats;
     }
+
+    /**
+     * Create a Tags entity object.
+     *
+     * @return Tags
+     */
+    public function createTags(): Tags
+    {
+        $class = $this->getEntityClass(Tags::class);
+        return new $class();
+    }
+
+    /**
+     * Get the row associated with a specific tag string.
+     *
+     * @param string $tag       Tag to look up.
+     * @param bool   $create    Should we create the row if it does not exist?
+     * @param bool   $firstOnly Should we return the first matching row (true)
+     * or the entire result set (in case of multiple matches)?
+     *
+     * @return mixed Matching row/result set if found or created, null otherwise.
+     */
+    public function getByText($tag, $create = true, $firstOnly = true)
+    {
+        $dql = 'SELECT t '
+            . 'FROM ' . $this->getEntityClass(Tags::class) . ' t '
+            . 'WHERE ' . ($this->caseSensitive ? 't.tag = :tag' : 'LOWER(t.tag) = LOWER(:tag) ');
+        $parameters = compact('tag');
+        $query = $this->entityManager->createQuery($dql);
+        $query->setParameters($parameters);
+        $result =  $query->getResult();
+        if (count($result) == 0 && $create) {
+            $row = $this->createTags()
+                ->setTag($this->caseSensitive ? $tag : mb_strtolower($tag, 'UTF8'));
+            try {
+                $this->persistEntity($row);
+            } catch (\Exception $e) {
+                $this->logError('Could not save tag: ' . $e->getMessage());
+                return false;
+            }
+            return $firstOnly ? $row : [$row];
+        }
+        return $firstOnly ? current($result) : $result;
+    }
+
+    /**
+     * Get a list of tags based on a sort method ($sort) and a where clause.
+     *
+     * @param string $sort  Sort/search parameter
+     * @param int    $limit Maximum number of tags (default = 100,
+     *                      < 1 = no limit)
+     * @param string $where Extra code to modify query (null for none)
+     * @param string $text  Tag to look up.
+     *
+     * @return array Tag details.
+     */
+    public function getTagList($sort = 'alphabetical', $limit = 100, $where = null, $text = null)
+    {
+        $tagClause = $this->caseSensitive ? 't.tag' : 'LOWER(t.tag)';
+        $dql = 'SELECT t.id as id, COUNT(DISTINCT(rt.resource)) as cnt, MAX(rt.posted) as posted '
+            . $tagClause . ' AS tag '
+            . 'FROM ' . $this->getEntityClass(ResourceTags::class) . ' rt '
+            . 'JOIN rt.tag t ';
+        $parameters = [];
+        switch ($where) {
+            case 'matchText':
+                $dql .= 'WHERE LOWER(t.tag) LIKE :text AND rt.resource is NOT NULL ';
+                $parameters['text'] = $text . '%';
+                break;
+            case 'discardUserList':
+                $dql .= 'WHERE rt.resource is NOT NULL ';
+                break;
+        }
+        $dql .= 'GROUP BY t.id, t.tag ';
+        switch ($sort) {
```

Comment:
Now that we're requiring PHP 8, you could simplify this switch into a [match expression](https://www.php.net/manual/en/control-structures.match.php) to improve readability.